src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,160 @@
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (checkIndex(index)) {
+            values[index] = value;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Error set"
+                    + System.lineSeparator()
+                    + "Size = " + localSize
+                    + " but index = " + index);
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (checkIndex(index)) {
+            localSize--;
+            Object[] newValues = new Object[values.length];
+            System.arraycopy(values, 0, newValues, 0, index);
+            System.arraycopy(values, index + 1, newValues, index, localSize - index);
+            T obj = (T) values[index];
+            values = newValues;
+            return obj;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Error remove by index"
+                    + System.lineSeparator()
+                    + "Size = " + localSize
+                    + " but index = " + index);
+        }
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = getIndex(element);
+        if (index == -1) {
+            throw new NoSuchElementException("Error remove by value"
+                    + System.lineSeparator()
+                    + "Size = " + localSize
+                    + " but index = " + index);
+        } else {
+            return remove(index);
+        }
     }
 
     @Override
     public int size() {
-        return 0;
+        return localSize;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return localSize == 0;
+    }
+
+    private Object[] grow(int srcPos, int destPos, int length) {
+        Object[] newValues = new Object[values.length + DEFAULT_SIZE / 2];
+        System.arraycopy(values, srcPos, newValues, destPos, length);
+        return newValues;
+    }
+
+    private Object[] grow(int sizeCollection) {
+        int newLength = values.length + DEFAULT_SIZE / 2;
+        Object[] newValues = new Object[newLength];
+        System.arraycopy(values, 0, newValues, 0, localSize);
+        return newValues;
+    }
+
+    private Object[] insertAdd(int index) {
+        int rightDistance = localSize - index;
+        int newLength = localSize + DEFAULT_SIZE / 2;
+        Object[] newValues = new Object[newLength];
+        System.arraycopy(values, 0, newValues, 0, index);
+        System.arraycopy(values, index, newValues, index + 1, rightDistance);
+        return newValues;
+    }
+
+    private int getIndex(T value) {
+        for (int i = 0; i < localSize; i++) {
+            if (value == values[i] || value != null && value.equals(values[i])) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    private boolean checkIndex(int index) {
+        return index < localSize && index >= 0;
```

Comment:
```suggestion
    private void checkIndex(int index) {
        if (index >= localSize || index < 0) {
            throw new ArrayListIndexOutOfBoundsException("Error remove by index"
                  + System.lineSeparator()
                  + "Size = " + localSize
                  + " but index = " + index);
        }
```
викидання ексепшина теж краще винести